Stabilize group router spec

johanandren opened the pull request:
Resending if messages are lost. Also a general cleanup making it a bit easier to follow.

References #30671

johanandren committed:
Stabilize group router spec #30671

octonato commented on akka-cluster-typed/src/test/scala/akka/cluster/typed/GroupRouterSpec.scala:
```diff
@@ -31,153 +35,166 @@ object GroupRouterSpec {
+              Behaviors.same
+            }
+          case RetryPinging =>
+            // delivery isn't guaranteed so retry sending if we didn't get resonses in a timely fashion
```
```suggestion
            // delivery isn't guaranteed so retry sending if we didn't get responses in a timely fashion
```

octonato committed:
fix typo

octonato reviewed (approved):
LGTM

(spotted a typo that I fixed myself)

octonato commented:
failed because of https://github.com/akka/akka/issues/30664, I will restart the build

octonato committed:
formatting

octonato commented:
one more commit because of scalafmt

patriknw commented on akka-cluster-typed/src/test/scala/akka/cluster/typed/GroupRouterSpec.scala:
```diff
@@ -31,154 +35,165 @@ object GroupRouterSpec {
+            // the requested number of workers are started and registered with the receptionist
+            // start pinging them
+            ctx.log.debug("Saw {} workers, starting pinger", update.size)
+            ctx.spawn(Pinger(router, settings.messageCount, resultProbe.ref), "pinger")
```
If the setup is stable at this point, all workers registered and such, then I don't understand why we would need resend attempts in the Pinger? Messages are not lost under normal conditions.

johanandren replied:
I thought I saw `Upstream failed, cause: StreamTcpException` but now I see that is a different test (RemoteAskSpec) in the same log dump. Hmmm.

johanandren committed:
No retrying, but found a potential race condition and fixed it

johanandren commented:
Removed the retrying but figured out a potential race condition that is now fixed:

The test would create a router, subscribe to the receptionist and only start pinging once it saw an update from the receptionist with the right number of workers. This did not guarantee the router had seen the right number of workers yet. Changed to create the router only after observing the number of workers so that it will see all from start (first listing it gets from receptionist).

patriknw reviewed (approved):
LGTM
